File: bindings/wasm/satox_sdk.js
/**
 * @file satox_sdk.js
 * @brief JavaScript wrapper for Satox SDK WASM binding
 * @copyright Copyright (c) 2025 Satoxcoin Core Developers
 * @license MIT License
 */

// Default configuration
const DEFAULT_CONFIG = {
    data_dir: "./data",
    network: "mainnet",
    enable_mining: false,
    enable_sync: true,
    sync_interval_ms: 1000,
    mining_threads: 1,
    max_connections: 10,
    rpc_port: 7777,
    p2p_port: 67777,
    enable_rpc: true,
    enable_p2p: true,
    log_level: "info",
    log_file: "./logs/satox-sdk.log"
};

// SatoxSDK class
class SatoxSDK {
    constructor(config = {}) {
        this.config = { ...DEFAULT_CONFIG, ...config };
        this.isRunning = false;
        this.managers = {};
        this.wasmInstance = null;
    }

    /**
     * Initialize the SDK with WASM
     */
    async initialize() {
        if (this.isRunning) {
            throw new Error("SDK is already running");
        }

        try {
            // Load WASM module
            const response = await fetch('./satox_sdk.wasm');
            const wasmBuffer = await response.arrayBuffer();
            
            // Initialize WASM
            const wasmModule = await WebAssembly.instantiate(wasmBuffer, {
                env: {
                    memory: new WebAssembly.Memory({ initial: 256 }),
                    abort: () => console.error('WASM abort called'),
                    // Add other required environment functions
                }
            });

            this.wasmInstance = wasmModule.instance;
            
            // Initialize core SDK
            const result = this.wasmInstance.exports.satox_sdk_initialize();
            if (result !== 0) {
                throw new Error("Failed to initialize SDK");
            }

            // Create and initialize managers
            this.managers.core = this.wasmInstance.exports.satox_core_manager_create();
            this.managers.wallet = this.wasmInstance.exports.satox_wallet_manager_create();
            this.managers.security = this.wasmInstance.exports.satox_security_manager_create();
            this.managers.asset = this.wasmInstance.exports.satox_asset_manager_create();
            this.managers.nft = this.wasmInstance.exports.satox_nft_manager_create();
            this.managers.blockchain = this.wasmInstance.exports.satox_blockchain_manager_create();
            this.managers.ipfs = this.wasmInstance.exports.satox_ipfs_manager_create();
            this.managers.network = this.wasmInstance.exports.satox_network_manager_create();
            this.managers.database = this.wasmInstance.exports.satox_database_manager_create();

            // Initialize all managers
            const configJson = JSON.stringify(this.config);
            const configPtr = this.stringToPtr(configJson);

            for (const [name, manager] of Object.entries(this.managers)) {
                const initFunc = this.wasmInstance.exports[`satox_${name}_manager_initialize`];
                const result = initFunc(manager, configPtr);
                if (result !== 0) {
                    throw new Error(`Failed to initialize ${name} manager`);
                }
            }

            this.isRunning = true;
        } catch (error) {
            throw new Error(`SDK initialization failed: ${error.message}`);
        }
    }

    /**
     * Start all SDK services
     */
    start() {
        if (!this.isRunning) {
            throw new Error("SDK is not initialized");
        }

        for (const [name, manager] of Object.entries(this.managers)) {
            const startFunc = this.wasmInstance.exports[`satox_${name}_manager_start`];
            const result = startFunc(manager);
            if (result !== 0) {
                throw new Error(`Failed to start ${name} manager`);
            }
        }
    }

    /**
     * Stop all SDK services
     */
    stop() {
        if (!this.isRunning) {
            return;
        }

        for (const [name, manager] of Object.entries(this.managers)) {
            const stopFunc = this.wasmInstance.exports[`satox_${name}_manager_stop`];
            stopFunc(manager);
        }

        this.isRunning = false;
    }

    /**
     * Connect to the blockchain network
     */
    connect() {
        if (!this.isRunning) {
            throw new Error("SDK is not running");
        }
        const result = this.wasmInstance.exports.satox_network_manager_connect(this.managers.network);
        if (result !== 0) {
            throw new Error("Failed to connect to network");
        }
    }

    /**
     * Disconnect from the blockchain network
     */
    disconnect() {
        if (!this.isRunning) {
            return;
        }
        this.wasmInstance.exports.satox_network_manager_disconnect(this.managers.network);
    }

    /**
     * Shutdown the SDK
     */
    shutdown() {
        this.stop();
        this.wasmInstance.exports.satox_sdk_shutdown();
    }

    /**
     * Get manager by name
     */
    getManager(name) {
        return this.managers[name];
    }

    /**
     * Get SDK version
     */
    getVersion() {
        const versionPtr = this.wasmInstance.exports.satox_sdk_get_version();
        return this.ptrToString(versionPtr);
    }

    /**
     * Get configuration
     */
    getConfig() {
        return this.config;
    }

    /**
     * Check if running
     */
    isRunning() {
        return this.isRunning;
    }

    /**
     * Convert JavaScript string to WASM pointer
     */
    stringToPtr(str) {
        const encoder = new TextEncoder();
        const bytes = encoder.encode(str + '\0');
        const ptr = this.wasmInstance.exports.malloc(bytes.length);
        const memory = new Uint8Array(this.wasmInstance.exports.memory.buffer);
        memory.set(bytes, ptr);
        return ptr;
    }

    /**
     * Convert WASM pointer to JavaScript string
     */
    ptrToString(ptr) {
        if (ptr === 0) return null;
        
        const memory = new Uint8Array(this.wasmInstance.exports.memory.buffer);
        let end = ptr;
        while (memory[end] !== 0) end++;
        
        const bytes = memory.slice(ptr, end);
        const decoder = new TextDecoder();
        return decoder.decode(bytes);
    }

    /**
     * Create wallet
     */
    createWallet(name) {
        if (!this.isRunning) {
            throw new Error("SDK is not running");
        }
        const namePtr = this.stringToPtr(name);
        const resultPtr = this.wasmInstance.exports.satox_wallet_manager_create_wallet(this.managers.wallet, namePtr);
        return this.ptrToString(resultPtr);
    }

    /**
     * Get wallet info
     */
    getWalletInfo(address) {
        if (!this.isRunning) {
            throw new Error("SDK is not running");
        }
        const addressPtr = this.stringToPtr(address);
        const resultPtr = this.wasmInstance.exports.satox_wallet_manager_get_wallet_info(this.managers.wallet, addressPtr);
        return this.ptrToString(resultPtr);
    }

    /**
     * Generate keypair
     */
    generateKeypair() {
        if (!this.isRunning) {
            throw new Error("SDK is not running");
        }
        const resultPtr = this.wasmInstance.exports.satox_security_manager_generate_keypair(this.managers.security);
        return this.ptrToString(resultPtr);
    }

    /**
     * Sign data
     */
    signData(data, privateKey) {
        if (!this.isRunning) {
            throw new Error("SDK is not running");
        }
        const dataPtr = this.stringToPtr(data);
        const keyPtr = this.stringToPtr(privateKey);
        const resultPtr = this.wasmInstance.exports.satox_security_manager_sign_data(this.managers.security, dataPtr, keyPtr);
        return this.ptrToString(resultPtr);
    }

    /**
     * Verify signature
     */
    verifySignature(data, signature, publicKey) {
        if (!this.isRunning) {
            throw new Error("SDK is not running");
        }
        const dataPtr = this.stringToPtr(data);
        const sigPtr = this.stringToPtr(signature);
        const keyPtr = this.stringToPtr(publicKey);
        const result = this.wasmInstance.exports.satox_security_manager_verify_signature(this.managers.security, dataPtr, sigPtr, keyPtr);
        return result === 1;
    }

    /**
     * Create asset
     */
    createAsset(config, ownerAddress) {
        if (!this.isRunning) {
            throw new Error("SDK is not running");
        }
        const configPtr = this.stringToPtr(JSON.stringify(config));
        const addressPtr = this.stringToPtr(ownerAddress);
        const resultPtr = this.wasmInstance.exports.satox_asset_manager_create_asset(this.managers.asset, configPtr, addressPtr);
        return this.ptrToString(resultPtr);
    }

    /**
     * Get asset info
     */
    getAssetInfo(assetId) {
        if (!this.isRunning) {
            throw new Error("SDK is not running");
        }
        const assetPtr = this.stringToPtr(assetId);
        const resultPtr = this.wasmInstance.exports.satox_asset_manager_get_asset_info(this.managers.asset, assetPtr);
        return this.ptrToString(resultPtr);
    }

    /**
     * Create NFT
     */
    createNFT(assetId, metadata, ownerAddress) {
        if (!this.isRunning) {
            throw new Error("SDK is not running");
        }
        const assetPtr = this.stringToPtr(assetId);
        const metadataPtr = this.stringToPtr(JSON.stringify(metadata));
        const addressPtr = this.stringToPtr(ownerAddress);
        const resultPtr = this.wasmInstance.exports.satox_nft_manager_create_nft(this.managers.nft, assetPtr, metadataPtr, addressPtr);
        return this.ptrToString(resultPtr);
    }

    /**
     * Get NFT info
     */
    getNFTInfo(nftId) {
        if (!this.isRunning) {
            throw new Error("SDK is not running");
        }
        const nftPtr = this.stringToPtr(nftId);
        const resultPtr = this.wasmInstance.exports.satox_nft_manager_get_nft_info(this.managers.nft, nftPtr);
        return this.ptrToString(resultPtr);
    }

    /**
     * Get block info
     */
    getBlockInfo(hash) {
        if (!this.isRunning) {
            throw new Error("SDK is not running");
        }
        const hashPtr = this.stringToPtr(hash);
        const resultPtr = this.wasmInstance.exports.satox_blockchain_manager_get_block_info(this.managers.blockchain, hashPtr);
        return this.ptrToString(resultPtr);
    }

    /**
     * Get transaction info
     */
    getTransactionInfo(txid) {
        if (!this.isRunning) {
            throw new Error("SDK is not running");
        }
        const txidPtr = this.stringToPtr(txid);
        const resultPtr = this.wasmInstance.exports.satox_blockchain_manager_get_transaction_info(this.managers.blockchain, txidPtr);
        return this.ptrToString(resultPtr);
    }

    /**
     * Upload file to IPFS
     */
    uploadFile(filePath) {
        if (!this.isRunning) {
            throw new Error("SDK is not running");
        }
        const pathPtr = this.stringToPtr(filePath);
        const resultPtr = this.wasmInstance.exports.satox_ipfs_manager_upload_file(this.managers.ipfs, pathPtr);
        return this.ptrToString(resultPtr);
    }

    /**
     * Get IPFS file info
     */
    getIPFSFileInfo(hash) {
        if (!this.isRunning) {
            throw new Error("SDK is not running");
        }
        const hashPtr = this.stringToPtr(hash);
        const resultPtr = this.wasmInstance.exports.satox_ipfs_manager_get_file_info(this.managers.ipfs, hashPtr);
        return this.ptrToString(resultPtr);
    }

    /**
     * Create database
     */
    createDatabase(name) {
        if (!this.isRunning) {
            throw new Error("SDK is not running");
        }
        const namePtr = this.stringToPtr(name);
        const resultPtr = this.wasmInstance.exports.satox_database_manager_create_database(this.managers.database, namePtr);
        return this.ptrToString(resultPtr);
    }

    /**
     * List databases
     */
    listDatabases() {
        if (!this.isRunning) {
            throw new Error("SDK is not running");
        }
        const resultPtr = this.wasmInstance.exports.satox_database_manager_list_databases(this.managers.database);
        return this.ptrToString(resultPtr);
    }
}

// Module exports
if (typeof module !== 'undefined' && module.exports) {
    module.exports = SatoxSDK;
} else if (typeof window !== 'undefined') {
    window.SatoxSDK = SatoxSDK;
}

// Version information
SatoxSDK.VERSION = "1.0.0";
SatoxSDK.BUILD_DATE = "2025-06-30";
SatoxSDK.DEFAULT_CONFIG = DEFAULT_CONFIG; 

File: bindings/lua/satox_sdk.lua
--[[
 * @file satox_sdk.lua
 * @brief Main Lua binding for the Satox SDK
 * @copyright Copyright (c) 2025 Satoxcoin Core Developers
 * @license MIT License
]]

local ffi = require("ffi")
local json = require("cjson")

-- Load the native library
local satox_lib = ffi.load("satox-sdk")

-- FFI declarations
ffi.cdef[[
    // Core SDK functions
    int satox_sdk_initialize();
    void satox_sdk_shutdown();
    const char* satox_sdk_get_version();
    int satox_sdk_connect();
    int satox_sdk_disconnect();
    
    // Core manager functions
    void* satox_core_manager_create();
    void satox_core_manager_destroy(void* handle);
    int satox_core_manager_initialize(void* handle, const char* config_json);
    int satox_core_manager_start(void* handle);
    int satox_core_manager_stop(void* handle);
    
    // Wallet manager functions
    void* satox_wallet_manager_create();
    void satox_wallet_manager_destroy(void* handle);
    int satox_wallet_manager_initialize(void* handle, const char* config_json);
    int satox_wallet_manager_start(void* handle);
    int satox_wallet_manager_stop(void* handle);
    char* satox_wallet_manager_create_wallet(void* handle, const char* name);
    char* satox_wallet_manager_get_wallet_info(void* handle, const char* address);
    
    // Security manager functions
    void* satox_security_manager_create();
    void satox_security_manager_destroy(void* handle);
    int satox_security_manager_initialize(void* handle, const char* config_json);
    int satox_security_manager_start(void* handle);
    int satox_security_manager_stop(void* handle);
    char* satox_security_manager_generate_keypair(void* handle);
    char* satox_security_manager_sign_data(void* handle, const char* data, const char* private_key);
    int satox_security_manager_verify_signature(void* handle, const char* data, const char* signature, const char* public_key);
    
    // Asset manager functions
    void* satox_asset_manager_create();
    void satox_asset_manager_destroy(void* handle);
    int satox_asset_manager_initialize(void* handle, const char* config_json);
    int satox_asset_manager_start(void* handle);
    int satox_asset_manager_stop(void* handle);
    char* satox_asset_manager_create_asset(void* handle, const char* config_json, const char* owner_address);
    char* satox_asset_manager_get_asset_info(void* handle, const char* asset_id);
    
    // NFT manager functions
    void* satox_nft_manager_create();
    void satox_nft_manager_destroy(void* handle);
    int satox_nft_manager_initialize(void* handle, const char* config_json);
    int satox_nft_manager_start(void* handle);
    int satox_nft_manager_stop(void* handle);
    char* satox_nft_manager_create_nft(void* handle, const char* asset_id, const char* metadata_json, const char* owner_address);
    char* satox_nft_manager_get_nft_info(void* handle, const char* nft_id);
    
    // Blockchain manager functions
    void* satox_blockchain_manager_create();
    void satox_blockchain_manager_destroy(void* handle);
    int satox_blockchain_manager_initialize(void* handle, const char* config_json);
    int satox_blockchain_manager_start(void* handle);
    int satox_blockchain_manager_stop(void* handle);
    char* satox_blockchain_manager_get_block_info(void* handle, const char* hash);
    char* satox_blockchain_manager_get_transaction_info(void* handle, const char* txid);
    
    // IPFS manager functions
    void* satox_ipfs_manager_create();
    void satox_ipfs_manager_destroy(void* handle);
    int satox_ipfs_manager_initialize(void* handle, const char* config_json);
    int satox_ipfs_manager_start(void* handle);
    int satox_ipfs_manager_stop(void* handle);
    char* satox_ipfs_manager_upload_file(void* handle, const char* file_path);
    char* satox_ipfs_manager_get_file_info(void* handle, const char* hash);
    
    // Network manager functions
    void* satox_network_manager_create();
    void satox_network_manager_destroy(void* handle);
    int satox_network_manager_initialize(void* handle, const char* config_json);
    int satox_network_manager_start(void* handle);
    int satox_network_manager_stop(void* handle);
    int satox_network_manager_connect(void* handle);
    int satox_network_manager_disconnect(void* handle);
    
    // Database manager functions
    void* satox_database_manager_create();
    void satox_database_manager_destroy(void* handle);
    int satox_database_manager_initialize(void* handle, const char* config_json);
    int satox_database_manager_start(void* handle);
    int satox_database_manager_stop(void* handle);
    char* satox_database_manager_create_database(void* handle, const char* name);
    char* satox_database_manager_list_databases(void* handle);
]]

-- Utility function to convert C string to Lua string
local function cstring_to_lua(cstr)
    if cstr == nil then
        return nil
    end
    return ffi.string(cstr)
end

-- Utility function to convert Lua string to C string
local function lua_to_cstring(str)
    if str == nil then
        return nil
    end
    return ffi.new("char[?]", #str + 1, str)
end

-- Default configuration
local DEFAULT_CONFIG = {
    data_dir = "./data",
    network = "mainnet",
    enable_mining = false,
    enable_sync = true,
    sync_interval_ms = 1000,
    mining_threads = 1,
    max_connections = 10,
    rpc_port = 7777,
    p2p_port = 67777,
    enable_rpc = true,
    enable_p2p = true,
    log_level = "info",
    log_file = "./logs/satox-sdk.log"
}

-- SatoxSDK class
local SatoxSDK = {}
SatoxSDK.__index = SatoxSDK

-- Constructor
function SatoxSDK.new(config)
    local self = setmetatable({}, SatoxSDK)
    self.config = config or DEFAULT_CONFIG
    self.is_running = false
    self.managers = {}
    return self
end

-- Initialize the SDK
function SatoxSDK:initialize()
    if self.is_running then
        error("SDK is already running")
    end
    
    -- Initialize core SDK
    local result = satox_lib.satox_sdk_initialize()
    if result ~= 0 then
        error("Failed to initialize SDK")
    end
    
    -- Create and initialize managers
    self.managers.core = satox_lib.satox_core_manager_create()
    self.managers.wallet = satox_lib.satox_wallet_manager_create()
    self.managers.security = satox_lib.satox_security_manager_create()
    self.managers.asset = satox_lib.satox_asset_manager_create()
    self.managers.nft = satox_lib.satox_nft_manager_create()
    self.managers.blockchain = satox_lib.satox_blockchain_manager_create()
    self.managers.ipfs = satox_lib.satox_ipfs_manager_create()
    self.managers.network = satox_lib.satox_network_manager_create()
    self.managers.database = satox_lib.satox_database_manager_create()
    
    -- Initialize all managers
    local config_json = json.encode(self.config)
    local config_cstr = lua_to_cstring(config_json)
    
    for name, manager in pairs(self.managers) do
        local init_func = satox_lib["satox_" .. name .. "_manager_initialize"]
        local result = init_func(manager, config_cstr)
        if result ~= 0 then
            error("Failed to initialize " .. name .. " manager")
        end
    end
    
    self.is_running = true
end

-- Start all services
function SatoxSDK:start()
    if not self.is_running then
        error("SDK is not initialized")
    end
    
    for name, manager in pairs(self.managers) do
        local start_func = satox_lib["satox_" .. name .. "_manager_start"]
        local result = start_func(manager)
        if result ~= 0 then
            error("Failed to start " .. name .. " manager")
        end
    end
end

-- Stop all services
function SatoxSDK:stop()
    if not self.is_running then
        return
    end
    
    for name, manager in pairs(self.managers) do
        local stop_func = satox_lib["satox_" .. name .. "_manager_stop"]
        stop_func(manager)
    end
    
    self.is_running = false
end

-- Connect to network
function SatoxSDK:connect()
    if not self.is_running then
        error("SDK is not running")
    end
    local result = satox_lib.satox_network_manager_connect(self.managers.network)
    if result ~= 0 then
        error("Failed to connect to network")
    end
end

-- Disconnect from network
function SatoxSDK:disconnect()
    if not self.is_running then
        return
    end
    satox_lib.satox_network_manager_disconnect(self.managers.network)
end

-- Shutdown the SDK
function SatoxSDK:shutdown()
    self:stop()
    satox_lib.satox_sdk_shutdown()
end

-- Get manager by name
function SatoxSDK:get_manager(name)
    return self.managers[name]
end

-- Get version
function SatoxSDK:get_version()
    local version_ptr = satox_lib.satox_sdk_get_version()
    return cstring_to_lua(version_ptr)
end

-- Get configuration
function SatoxSDK:get_config()
    return self.config
end

-- Check if running
function SatoxSDK:is_running()
    return self.is_running
end

-- Module exports
local M = {}

-- Create new SDK instance
function M.new(config)
    return SatoxSDK.new(config)
end

-- Get SDK version
function M.get_version()
    local version_ptr = satox_lib.satox_sdk_get_version()
    return cstring_to_lua(version_ptr)
end

-- Initialize SDK (global)
function M.initialize()
    local result = satox_lib.satox_sdk_initialize()
    if result ~= 0 then
        error("Failed to initialize SDK")
    end
end

-- Shutdown SDK (global)
function M.shutdown()
    satox_lib.satox_sdk_shutdown()
end

-- Default configuration
M.DEFAULT_CONFIG = DEFAULT_CONFIG

-- Version information
M.VERSION = "1.0.0"
M.BUILD_DATE = "2025-06-30"

return M 